app/coffee/AuthHandler.coffee

```diff
@@ -0,0 +1,85 @@
+_ = require("lodash")
+bcrypt = require("bcrypt")
+crypto = require("crypto")
+logger = require("logger-sharelatex")
+
+{ db } = require("./MongoHandler")
+Errors = require("./Errors")
+
+TOKEN_TTL_MS = 60 * 60 * 1000    # 1 hour
+TOKEN_SEARCH_PREFIX_LENGTH = 16  # 8 bytes
+
+module.exports = AuthHandler =
+	login: (email, password, callback) ->
+		db.users.findOne {email: email.toLowerCase()}, (err, user) ->
+			if err?
+				return callback(err)
+			if !user
+				return callback(new Errors.AuthenticationError("user not found"))
+			if !user.hashedPassword
+				return callback(new Errors.AuthenticationError("user does not have a password"))
+			bcrypt.compare password, user.hashedPassword, (err, match) ->
+				if err?
+					return callback(err)
+				if !match
+					return callback(new Errors.AuthenticationError("incorrect password"))
+				callback(null, user._id.toString())
+
+	oneTimeLogin: (email, token, callback) ->
+		now = new Date()
+		normalizedEmail = email.toLowerCase()
+		tokenPrefix = token.slice(0, TOKEN_SEARCH_PREFIX_LENGTH)
+		tokenPrefixRegexp = new RegExp("^#{_.escapeRegExp(tokenPrefix)}")
+		query = {
+			email: normalizedEmail
+			token: tokenPrefixRegexp
+			expiresAt: { $gt: now }
+			usedAt: { $exists: false }
+		}
+		update = {
+			$set: { usedAt: now }
```

Comment:
This will optimistically set `usedAt` even if the comparison on the latter half of the token fails. That's probably OK, but it is a bit odd. Also, what happens if you pass an empty token?

Reply:
Right. I'll do it in two steps.

If you pass an empty token, it will get you one of your tokens and fail at the comparison step. I'll add a test for it.